src/dimension_slice.c

```diff
@@ -337,3 +385,141 @@ dimension_slice_insert_multi(DimensionSlice **slices, Size num_slices)
 
 	heap_close(rel, RowExclusiveLock);
 }
+
+static Datum
+create_range_datum(FunctionCallInfo fcinfo, DimensionSlice *slice)
+{
+	TupleDesc	tupdesc;
+	Datum		values[2];
+	bool		nulls[2] = {false};
+	HeapTuple	tuple;
+
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		elog(ERROR, "Function returning record called in context that cannot accept type record");
+
+	tupdesc = BlessTupleDesc(tupdesc);
+
+	values[0] = Int64GetDatum(slice->fd.range_start);
+	values[1] = Int64GetDatum(slice->fd.range_end);
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+
+	return HeapTupleGetDatum(tuple);
+}
+
+static DimensionSlice *
+calculate_open_range_default(Dimension *dim, int64 value)
```

Comment:
This entire code path is better conceived of as part of the constructor of a DimensionSlice rather than a method on dimension, IMO. This can be seen by the fact that the logic touches upon a lot of implementation details of ranges and only extracts the dimension id from the dimension. Do we have a convention for constructors? Should I replace the Dimension argument with dimension_id instead?

Reply:
As it stands, I cannot see a proper diff of what has changed, which is a problem in itself. Further, this is an operation on a Dimension object (which defines how a hyperspace is spanned and split into ranges in one dimension). That this function is a calculation on a dimension object can be seen from the fact that it (and the related function) uses `interval_length`, `num_slices` and `dimension_id`, all attributes that define the Dimension and tells us how it is divided into slices. Just passing dimension ID to this function will not be sufficient. The only slice-related info I can see being used is the MAX value, which actually need not be defined as part of dimension_slice, although it is fine with me. Slices themselves need not know how their ranges were calculated. There is nothing strange that operations on one object can produce another object, as in this case.

Regarding this being a constructor; you can see that the actual dimension slice constructor is already called at the end of this function. And that constructor is in `dimension_slice.c` already.

In any case, I think this need not be part of this PR and just makes things harder to review and makes us have this discussion instead of focusing on the stuff that is important for fixing the bug.